From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: 404Setup <153366651+404Setup@users.noreply.github.com>
Date: Sat, 7 Sep 2024 14:59:07 +0800
Subject: [PATCH] Paper PR: Fix fetching biomes during world generation

Source in https://github.com/PaperMC/Paper/pull/11086

Fixes PaperMC/Paper #11084

getBiome() on LevelReader calls getNoiseBiome(), which itself calls getChunk() in order to get the biome cached inside the chunk.
However, calling getChunk() on WorldGenRegion will cause a crash if the chunk isn't found: "Requested chunk unavailable during world generation"
This means that features which attempt to fetch biomes during generation (LakeFeature/SnowAndFreezeFeature)
can cause the server to crash.
This patch ensures that these features are fetching biomes in a way that doesn't call getChunk()

diff --git a/src/main/java/net/minecraft/world/level/levelgen/feature/LakeFeature.java b/src/main/java/net/minecraft/world/level/levelgen/feature/LakeFeature.java
index 830558f0e5d3d18463a8348c629463ccc239ad65..227362dd23d73cef8bb94293435d3032971ecbcb 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/feature/LakeFeature.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/feature/LakeFeature.java
@@ -134,7 +134,7 @@ public class LakeFeature extends Feature<LakeFeature.Configuration> {
                     for (int ac = 0; ac < 16; ac++) {
                         int ad = 4;
                         BlockPos blockPos4 = blockPos.offset(ab, 4, ac);
-                        if (worldGenLevel.getBiome(blockPos4).value().shouldFreeze(worldGenLevel, blockPos4, false)
+                        if (worldGenLevel.getUncachedNoiseBiome(blockPos4.getX(), blockPos4.getY(), blockPos4.getZ()).value().shouldFreeze(worldGenLevel, blockPos4, false) // Paper - Fix fetching biomes during world generation
                             && this.canReplaceBlock(worldGenLevel.getBlockState(blockPos4))) {
                             worldGenLevel.setBlock(blockPos4, Blocks.ICE.defaultBlockState(), 2);
                         }
diff --git a/src/main/java/net/minecraft/world/level/levelgen/feature/SnowAndFreezeFeature.java b/src/main/java/net/minecraft/world/level/levelgen/feature/SnowAndFreezeFeature.java
index 3018026b7d11c8714d05dec4d131d3973c8f2398..659c8ccc8735c82e5c277ecb239f5615e5fd90c3 100644
--- a/src/main/java/net/minecraft/world/level/levelgen/feature/SnowAndFreezeFeature.java
+++ b/src/main/java/net/minecraft/world/level/levelgen/feature/SnowAndFreezeFeature.java
@@ -30,7 +30,7 @@ public class SnowAndFreezeFeature extends Feature<NoneFeatureConfiguration> {
                 int m = worldGenLevel.getHeight(Heightmap.Types.MOTION_BLOCKING, k, l);
                 mutableBlockPos.set(k, m, l);
                 mutableBlockPos2.set(mutableBlockPos).move(Direction.DOWN, 1);
-                Biome biome = worldGenLevel.getBiome(mutableBlockPos).value();
+                Biome biome = worldGenLevel.getUncachedNoiseBiome(mutableBlockPos.getX(), mutableBlockPos.getY(), mutableBlockPos.getZ()).value(); // Paper - Fix fetching biomes during world generation
                 if (biome.shouldFreeze(worldGenLevel, mutableBlockPos2, false)) {
                     worldGenLevel.setBlock(mutableBlockPos2, Blocks.ICE.defaultBlockState(), 2);
                 }
